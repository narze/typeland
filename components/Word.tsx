/** @jsx jsx */
import { css, jsx } from '@emotion/core'
import tw from '@tailwindcssinjs/macro'
import React from 'react'

const s = {
  correct: tw`
    text-green-600
  `,
  wrong: tw`
    text-red-400
  `,
  caret: [
    tw`
      h-4
      border border-blue-500
    `,
    css`
      width: 1px;
    `,
  ],
}

export interface WordProps {
  template: string
  userInput: string
  showCaret: boolean
}

export const Word: React.FC<WordProps> = ({
  template,
  userInput,
  showCaret,
}) => {
  return (
    <span>
      {showCaret && userInput.length == 0 ? (
        <span css={s.caret} data-testid="caret"></span>
      ) : null}

      {Array.from(
        Array(Math.min(template.length, userInput.length)).keys()
      ).map((i) => {
        const templateChar = template[i]
        const userInputChar = userInput[i]
        const displayChar = userInputChar || templateChar

        return (
          <React.Fragment key={i}>
            {templateChar == userInputChar ? (
              <span css={s.correct} className="correct" data-testid="correct">
                {displayChar}
              </span>
            ) : (
              <span css={s.wrong} className="wrong" data-testid="wrong">
                {displayChar}
              </span>
            )}
          </React.Fragment>
        )
      })}

      {showCaret && userInput.length != 0 ? (
        <span css={s.caret} data-testid="caret"></span>
      ) : null}
    </span>
  )
}
