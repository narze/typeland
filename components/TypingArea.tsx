/** @jsx jsx */
import { jsx } from '@emotion/core'
import tw from '@tailwindcssinjs/macro'
import { Word } from './Word'
import React, { useEffect } from 'react'

const s = {
  typingArea: tw`
    text-xl
    antialiased
  `,
  template: tw`
    my-1
    break-words
  `,
  user: tw`
    my-1
    break-words
  `,
}

export enum Mode {
  typealong = 'typealong',
}

export interface TypingAreaProps {
  words: Array<string>
  userWords: Array<string>
  showCaret: boolean
  mode?: Mode
  onStatsUpdate?: (stats: { correct: number }) => void
}

export const TypingArea: React.FC<TypingAreaProps> = ({
  words,
  userWords,
  showCaret,
  mode,
  onStatsUpdate,
}) => {
  useEffect(() => {
    if (onStatsUpdate) {
      const stats = {
        correct: 0,
        wrong: 0,
      }

      // TODO: Optimize so that userWords are not iterate every time
      userWords.forEach((text, i) => {
        if (text == words[i]) {
          stats.correct += 1
        } else {
          stats.wrong += 1
        }
      })

      onStatsUpdate(stats)
    }
  }, [userWords.length, (userWords[words.length - 1] || []).length])

  if (mode == 'typealong') {
    const remainingWords = words.slice(userWords.length)

    return (
      <div css={s.typingArea}>
        <span css={s.user} data-testid="user">
          {userWords.map((text, i) => {
            return (
              <React.Fragment key={i}>
                <Word
                  template={words[i]}
                  userInput={text}
                  showCaret={showCaret && i == userWords.length - 1}
                  mode={mode}
                />
                <span>&nbsp;</span>
              </React.Fragment>
            )
          })}
        </span>
        <span css={s.template} data-testid="template">
          {remainingWords.join(' ')}
        </span>
      </div>
    )
  }

  return (
    <div css={s.typingArea}>
      <p css={s.template} data-testid="template">
        {words.join(' ')}
      </p>
      <p css={s.user} data-testid="user">
        {userWords.map((text, i) => {
          return (
            <React.Fragment key={i}>
              <Word
                template={words[i]}
                userInput={text}
                showCaret={showCaret && i == userWords.length - 1}
              />
              <span>&nbsp;</span>
            </React.Fragment>
          )
        })}
      </p>
    </div>
  )
}
