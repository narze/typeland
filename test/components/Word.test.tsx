import React from 'react'
import { render } from '../testUtils'
import { Word, Mode } from '@/components/Word'

it('renders single word with caret', () => {
  const props = {
    template: 'doge',
    userInput: 'do',
    showCaret: true,
    mode: Mode.typealong,
  }

  const { asFragment, getAllByTestId } = render(<Word {...props} />, {})
  expect(
    getAllByTestId('correct')
      .map((e) => e.textContent)
      .join('')
  ).toBe('do')
  expect(
    getAllByTestId('pending')
      .map((e) => e.textContent)
      .join('')
  ).toBe('ge')

  const o = getAllByTestId('correct').slice(-1)[0]
    .nextElementSibling as HTMLElement
  expect(o.dataset.testid).toBe('caret')

  expect(asFragment()).toMatchSnapshot()
})
