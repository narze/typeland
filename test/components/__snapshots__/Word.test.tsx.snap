// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders single word with caret 1`] = `
<DocumentFragment>
  .emotion-5 {
  overflow-wrap: normal;
  word-break: normal;
}

.emotion-0 {
  --text-opacity: 1;
  color: #057a55;
  color: rgba(5,122,85,var(--text-opacity));
}

.emotion-2 {
  --bg-opacity: 1;
  display: inline-block;
  overflow-wrap: normal;
  word-break: normal;
  margin-bottom: -0.25rem;
  height: 1.5rem;
  background-color: #3f83f8;
  background-color: rgba(63,131,248,var(--bg-opacity));
  width: 2px;
  margin-left: -1px;
  margin-right: -1px;
}

.emotion-3 {
  --text-opacity: 1;
  color: #d2d6dc;
  color: rgba(210,214,220,var(--text-opacity));
}

<span
    class="emotion-5"
  >
    <span
      class="correct emotion-0"
      data-testid="correct"
    >
      d
    </span>
    <span
      class="correct emotion-0"
      data-testid="correct"
    >
      o
    </span>
    <span
      class="emotion-2"
      data-testid="caret"
    />
    <span
      class="pending emotion-3"
      data-testid="pending"
    >
      g
    </span>
    <span
      class="pending emotion-3"
      data-testid="pending"
    >
      e
    </span>
  </span>
</DocumentFragment>
`;

exports[`renders template character on wrong typing in typealong mode 1`] = `
<DocumentFragment>
  .emotion-6 {
  overflow-wrap: normal;
  word-break: normal;
}

.emotion-0 {
  --text-opacity: 1;
  color: #057a55;
  color: rgba(5,122,85,var(--text-opacity));
}

.emotion-1 {
  --text-opacity: 1;
  color: #f98080;
  color: rgba(249,128,128,var(--text-opacity));
}

.emotion-4 {
  --bg-opacity: 1;
  display: inline-block;
  overflow-wrap: normal;
  word-break: normal;
  margin-bottom: -0.25rem;
  height: 1.5rem;
  background-color: #3f83f8;
  background-color: rgba(63,131,248,var(--bg-opacity));
  width: 2px;
  margin-left: -1px;
  margin-right: -1px;
}

.emotion-5 {
  --text-opacity: 1;
  color: #d2d6dc;
  color: rgba(210,214,220,var(--text-opacity));
}

<span
    class="emotion-6"
  >
    <span
      class="correct emotion-0"
      data-testid="correct"
    >
      w
    </span>
    <span
      class="wrong emotion-1"
      data-testid="wrong"
    >
      o
    </span>
    <span
      class="wrong emotion-1"
      data-testid="wrong"
    >
      r
    </span>
    <span
      class="wrong emotion-1"
      data-testid="wrong"
    >
      l
    </span>
    <span
      class="emotion-4"
      data-testid="caret"
    />
    <span
      class="pending emotion-5"
      data-testid="pending"
    >
      d
    </span>
  </span>
</DocumentFragment>
`;

exports[`renders template character on wrong typing in typealong mode 2`] = `
<DocumentFragment>
  .emotion-8 {
  overflow-wrap: normal;
  word-break: normal;
}

.emotion-0 {
  --text-opacity: 1;
  color: #057a55;
  color: rgba(5,122,85,var(--text-opacity));
}

.emotion-1 {
  --text-opacity: 1;
  color: #f98080;
  color: rgba(249,128,128,var(--text-opacity));
}

.emotion-7 {
  --bg-opacity: 1;
  display: inline-block;
  overflow-wrap: normal;
  word-break: normal;
  margin-bottom: -0.25rem;
  height: 1.5rem;
  background-color: #3f83f8;
  background-color: rgba(63,131,248,var(--bg-opacity));
  width: 2px;
  margin-left: -1px;
  margin-right: -1px;
}

<span
    class="emotion-8"
  >
    <span
      class="correct emotion-0"
      data-testid="correct"
    >
      w
    </span>
    <span
      class="wrong emotion-1"
      data-testid="wrong"
    >
      o
    </span>
    <span
      class="wrong emotion-1"
      data-testid="wrong"
    >
      r
    </span>
    <span
      class="wrong emotion-1"
      data-testid="wrong"
    >
      l
    </span>
    <span
      class="wrong emotion-1"
      data-testid="wrong"
    >
      d
    </span>
    <span
      class="wrong emotion-1"
      data-testid="wrong"
    >
      x
    </span>
    <span
      class="wrong emotion-1"
      data-testid="wrong"
    >
      x
    </span>
    <span
      class="emotion-7"
      data-testid="caret"
    />
  </span>
</DocumentFragment>
`;

exports[`renders user character on wrong typing in default mode 1`] = `
<DocumentFragment>
  .emotion-5 {
  overflow-wrap: normal;
  word-break: normal;
}

.emotion-0 {
  --text-opacity: 1;
  color: #057a55;
  color: rgba(5,122,85,var(--text-opacity));
}

.emotion-1 {
  --text-opacity: 1;
  color: #f98080;
  color: rgba(249,128,128,var(--text-opacity));
}

.emotion-4 {
  --bg-opacity: 1;
  display: inline-block;
  overflow-wrap: normal;
  word-break: normal;
  margin-bottom: -0.25rem;
  height: 1.5rem;
  background-color: #3f83f8;
  background-color: rgba(63,131,248,var(--bg-opacity));
  width: 2px;
  margin-left: -1px;
  margin-right: -1px;
}

<span
    class="emotion-5"
  >
    <span
      class="correct emotion-0"
      data-testid="correct"
    >
      w
    </span>
    <span
      class="wrong emotion-1"
      data-testid="wrong"
    >
      w
    </span>
    <span
      class="wrong emotion-1"
      data-testid="wrong"
    >
      w
    </span>
    <span
      class="wrong emotion-1"
      data-testid="wrong"
    >
      w
    </span>
    <span
      class="emotion-4"
      data-testid="caret"
    />
  </span>
</DocumentFragment>
`;
